crossdock/client/errorstchout/behavior.go

```diff
@@ -0,0 +1,176 @@
+				assert.Equal(tchannel.ErrCodeBadRequest, code, "must produce bad request error")
+			},
+		},
+		// TODO test invalid headers
+		{
+			procedure: "",
+			body:      []byte{},
+			headers:   []byte("{}"),
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.False(isAppErr, "missing procedure must not produce an application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "missing procedure must produce system error")
+				if !ok {
+					return
+				}
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(tchannel.ErrCodeBadRequest, code, "missing procedure must produce bad request error")
+				assert.Contains(err.Error(), "missing procedure", "must mention missing procedure in error message")
+			},
+		},
+		{
+			procedure: "no-such-procedure",
+			body:      []byte{},
+			headers:   []byte("{}"),
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.False(isAppErr, "no-such-procedure must not produce application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "no-such-procedure must produce  system error")
+				if !ok {
+					return
+				}
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(tchannel.ErrCodeBadRequest, code, "must produce bad request error")
+				assert.Contains(err.Error(), `unrecognized procedure "no-such-procedure"`, "must mention unrecongized procedure in error message")
+			},
+		},
+		{
+			procedure: "bad-response",
+			body:      []byte("{}"),
+			headers:   []byte("{}"),
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.False(isAppErr, "bad-response must not produce an application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "bad-response must produce system error")
+				if !ok {
+					return
+				}
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(tchannel.ErrCodeUnexpected, code, "bad-response must produce unexpected error")
+				assert.Contains(err.Error(), `failed to encode "json"`, "must mention failure to encode JSON in error message")
+			},
+		},
+		{
+			procedure: "unexpected-error",
+			body:      []byte("{}"),
+			headers:   []byte("{}"),
+			validate: func(res3 []byte, isAppErr bool, err error) {
+				assert.False(isAppErr, "unexpected-error procedure must not produce application error")
+				err, ok := err.(tchannel.SystemError)
+				assert.True(ok, "unexpected-error procedure must produce system error")
+				code := tchannel.GetSystemErrorCode(err)
+				assert.Equal(tchannel.ErrCodeUnexpected, code, "must produce transport error")
+			},
+		},
+	}
+
+	server := t.Param(params.Server)
+	serverHostPort := fmt.Sprintf("%v:%v", server, serverPort)
+
+	ch, err := tchannel.NewChannel(serviceName, nil)
+	fatals.NoError(err, "Could not create channel")
+
+	peer := ch.Peers().Add(serverHostPort)
+
+	for _, tt := range tests {
+		(func(tt test) {
```

Comment:
If the deferred function does not have to run before the next test, I'd skip the inline function completely. In this case it looks like the only defer is for the `cancel` which is mostly a nice-to-have to release resources, it isn't actually required to run before the next test. Removing the inline function will reduce a bit of nesting.